streams/src/main/java/org/apache/kafka/streams/state/internals/BufferValue.java

```diff
@@ -16,83 +16,162 @@
  */
 package org.apache.kafka.streams.state.internals;
 
+import org.apache.kafka.streams.processor.internals.ProcessorRecordContext;
+
 import java.nio.ByteBuffer;
 import java.util.Arrays;
 import java.util.Objects;
 
 public final class BufferValue {
-    private final ContextualRecord record;
+    private static final int NULL_VALUE_SENTINEL = -1;
+    private static final int OLD_PREV_DUPLICATE_VALUE_SENTINEL = -2;
     private final byte[] priorValue;
+    private final byte[] oldValue;
+    private final byte[] newValue;
+    private final ProcessorRecordContext recordContext;
+
+    BufferValue(final byte[] priorValue, final byte[] oldValue, final byte[] newValue, final ProcessorRecordContext recordContext) {
+        this.oldValue = oldValue;
+        this.newValue = newValue;
+        this.recordContext = recordContext;
+
+        // This de-duplicates the prior and old references.
+        // If they were already the same reference, the comparison is trivially fast, so we don't specifically check
+        // for that case.
+        if (Arrays.equals(priorValue, oldValue)) {
+            this.priorValue = oldValue;
+        } else {
+            this.priorValue = priorValue;
+        }
+    }
 
-    BufferValue(final ContextualRecord record,
-                final byte[] priorValue) {
-        this.record = record;
-        this.priorValue = priorValue;
+    byte[] priorValue() {
+        return priorValue;
     }
 
-    ContextualRecord record() {
-        return record;
+    byte[] oldValue() {
+        return oldValue;
     }
 
-    byte[] priorValue() {
-        return priorValue;
+    byte[] newValue() {
+        return newValue;
+    }
+
+    ProcessorRecordContext context() {
+        return recordContext;
     }
 
     static BufferValue deserialize(final ByteBuffer buffer) {
-        final ContextualRecord record = ContextualRecord.deserialize(buffer);
+        final ProcessorRecordContext context = ProcessorRecordContext.deserialize(buffer);
+
+        final byte[] priorValue = extractValue(buffer);
+
+        final byte[] oldValue;
+        final int oldValueLength = buffer.getInt();
+        if (oldValueLength == NULL_VALUE_SENTINEL) {
+            oldValue = null;
+        } else if (oldValueLength == OLD_PREV_DUPLICATE_VALUE_SENTINEL) {
```

Comment:
We also de-duplicate in serial form by making use of the oldValue "length" field to store a sentinel indicating that the old value is the same as the prior value, so we don't need to serialize it twice.